pandas/core/frame.py

```diff
@@ -8559,6 +8580,54 @@ def blk_func(values):
                 out[:] = coerce_to_dtypes(out.values, df.dtypes)
             return out
 
+        def array_func(values):
+            if isinstance(values, ExtensionArray):
+                return values._reduce(name, skipna=skipna, **kwds)
+            else:
+                return op(values, skipna=skipna, **kwds)
+
+        # all other options with axis=0 are done column-array-wise
+        if axis == 0:
+
+            def _constructor(df, result, index=None):
+                index = index if index is not None else df.columns
+                if len(result):
+                    return df._constructor_sliced(result, index=index)
+                else:
+                    # set correct dtype for empty result
+                    dtype = "bool" if filter_type == "bool" else "float64"
+                    return df._constructor_sliced(result, index=index, dtype=dtype)
+
+            def _reduce_columns(df, op):
+                result = [op(arr) for arr in df._iter_column_arrays()]
+                return _constructor(df, result)
+
+            df = self
+            if numeric_only is True:
+                df = _get_data(axis_matters=True)
+
+            if numeric_only is not None:
+                return _reduce_columns(df, array_func)
+            else:
+                # need to catch and ignore exceptions when numeric_only=None
+                try:
+                    return _reduce_columns(df, array_func)
+                except TypeError:
+                    # if column-wise fails and numeric_only was None, we try
+                    # again but removing those columns for which it fails
+                    result = []
+                    indices = []
+                    for i, arr in enumerate(df._iter_column_arrays()):
+                        try:
+                            res = array_func(arr)
+                        except Exception:
```

Comment:
can this be any tighter?

Reply:
We use the same Exception in `frame_apply`:

https://github.com/pandas-dev/pandas/blob/69cc851a2ad077529c1a1fd4a3484a24e6c32b50/pandas/core/apply.py#L336-L344

Now, in apply this can be any user provided function, so needs to be general. Here, we know we are calling `EA._reduce` or one of the nanops ops, so I *suppose* it can be tighter. I assume eg nanops raise TypeError if it is a dtype they don't support?

Reply:
Makes sense, thanks.